substrate/frame/support/procedural/src/pallet/parse/config.rs

```diff
@@ -439,6 +496,47 @@ impl ConfigDef {
 				}
 			}
 
+			if let Some(span) = already_collected_associated_type {
+				// Events and constants are already propagated to the metadata
+				if is_event {
+					return Err(syn::Error::new(
+						span,
+						"Invalid #[pallet::include_metadata] for `type RuntimeEvent`. \
+						The associated type `RuntimeEvent` is already collected in the metadata.",
+					))
+				}
+
+				if already_constant {
+					return Err(syn::Error::new(
+						span,
+						"Invalid #[pallet::include_metadata]: conflict with #[pallet::constant]. \
+						Pallet constant already collect the metadata for the type.",
+					))
+				}
+
+				if let syn::TraitItem::Type(ref ty) = trait_item {
+					if !contains_type_info_bound(ty) {
```

Comment:
Could we implement here a mix of:

https://docs.rs/static_assertions/1.1.0/src/static_assertions/assert_impl.rs.html#329-356

and

https://doc.rust-lang.org/beta/reference/attributes/diagnostics.html#the-diagnosticon_unimplemented-attribute

This should give some very nice error message.